### PR TITLE
Update RE2

### PR DESCRIPTION
**Pull request by mppf**
* Upgrade RE2 to re2 commit 2b16c27
* Fix a trivial C++ compilation warning

I verified that the current diff with RE2 makes sense.

- [x] test/regexp with clang on Mac OS X
- [x] full local testing
- [x] update commit ID in README, make it look more like other READMEs in third-party
- [x] perf check regex-dna / regex-dna-redux: perf improves by 10% - 25%

```
16 trials
new regex-dna best: 3.19 worst: 3.26
new regex-dna-redux best: 3.80 worst: 3.84
old regex-dna best: 3.50 worst: 3.55
old regex-dna-redux best: 4.82 worst: 4.89
```

Reviewed by @ronawho - thanks!

### CODE REVIEW

**Comment by mppf** [comment]
@ronawho - could you sign off on this or let me know what other steps you think I need to take?

**Comment by ronawho** [comment]
https://github.com/chapel-lang/chapel/issues/5271 probably has some TODOs in it. 

I didn't look at this too carefully yet, but I see some cmake stuff. Has re2 switched from autoconf/configure to cmake? If so, does this mean cmake is now a dependency for us?

**Comment by mppf** [comment]
AFAIK RE2 is currently supporting both CMake and Make, so it's not a dependence. It seems that the code now assumes C++11 (but we're used to needing that for LLVM builds).

**Comment by ronawho** [comment]
llvm is always built with the host compiler though, right ? I'd bet intels C++11 support is good, but what about cray/pgi?

**Comment by ronawho** [comment]
Also, do you have any sense of what improvements this brings us? 

It looks like re2 doesn't have a ChangeLog or releases notes, at least not that I found quickly :(

**Comment by mppf** [comment]
I can't really describe what we're getting with the updates easily either. AFAIK the commit log is all we have. It's multiple years of updates though.

**Comment by ronawho** [comment]
I think it's worth running a regexp heavy benchmark just to check that this won't introduce any major regressions. 

When upgrading jemalloc I test binary-trees. For re2, I'd recommend testing regex-dna/regex-dna redux.

**Comment by mppf** [comment]
@ronawho - see the PR message for the perf test result. Is there anything else you think I need to do for this PR in particular?

File: third-party/re2/Makefile [hunk 1]
```diff
@@ -49,7 +49,7 @@ $(RE2_BUILD_SUBDIR):
 	mkdir -p $@
 
 $(RE2_H_FILE): $(RE2_BUILD_SUBDIR)
-	cd re2 && \
+	cd re2-src && \
```
Comment (ronawho) [inline third-party/re2/Makefile]: `cd $(RE2_SUBDIR) && \` ?

**Comment by ronawho** [comment]
re2 link in the README should be updated to the github repo. I also think it's worth including the "version" number in the README and maybe making it more uniform with the other third-party READMES (e.g. gasnet, jemalloc). Maybe something like:


```rst
=====================
re2 README for Chapel
=====================

This copy of re2 2017-04-01 (2b16c27) is being released with Chapel
for convenience and was obtained from:

  https://github.com/google/re2

Any Chapel issues that seem to be related to re2 should be directed
to the Chapel team at http://chapel.cray.com/bugs.html.


Chapel modifications to re2
===========================
We modify the RE2 library to allow regular expressions in the format
string for Chapel's readf() for searching on QIO channels.

- moved NFA and OnePass searches to be implemented with templates
  to allow string types other than const char*
  Also modified these to use begin_reading() instead of begin()
  for the pointer that will be dereferenced.
- added file_strings.h to represent a portion of a file accessible
  by calling some read routine (through a pointer)
  and modified string_piece to include it.
- implemented RE2::MatchFile which searches a buffer
  (optionally) and then some other string type.
- RE2 constructor now computes min/max possible match length
  for use in MatchFile.

```

**Pull request merged**